### PR TITLE
rbac: rules_stat_prefix to emit metrics with a prefix

### PR DESCRIPTION
**Pull request by thomasvnoort**
Commit Message: rbac: rules_stat_prefix to emit metrics with a prefix
Risk Level: low
Testing: updated existing tests
Docs Changes: done
Release Notes: done
Platform Specific Features: N/A

Additional Description: 
This is akin to shadow_rules_stat_prefix but for non-shadowing rules. Since only shadow rules emit dynamic metadata, this prefix only applies to metrics.

edit: This only applies to the http RBAC filter, not the network RBAC filter since that already supports `stat_prefix`.

### CODE REVIEW

File: api/envoy/extensions/filters/network/rbac/v3/rbac.proto
Comment (adisuissa) [inline api/envoy/extensions/filters/network/rbac/v3/rbac.proto:55]: I may be missing something, but what isn't the `stat_prefix` sufficient to distinguish between shadow/non-shadow?

Comment (thomasvnoort) [reply api/envoy/extensions/filters/network/rbac/v3/rbac.proto:55]: Do you mean `stat_prefix` as in `http.<stat_prefix>.rbac.`? That one is inherited from the HTTP connection manager and the same if you have multiple RBAC filters under one manager. 

> The [stat prefix](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto#envoy-v3-api-field-extensions-filters-network-http-connection-manager-v3-httpconnectionmanager-stat-prefix) comes from the owning HTTP connection manager.

`stat_prefix` is also the same for shadow/non-shadow rules, which is why `shadow_rules_stat_prefix` was introduced a while ago.

**Comment by thomasvnoort** [comment]
@adisuissa @yanavlasov Is there anything else you need from me before this PR can get reviewed? Thanks!

File: api/envoy/extensions/filters/network/rbac/v3/rbac.proto
Comment (adisuissa) [reply api/envoy/extensions/filters/network/rbac/v3/rbac.proto:55]: Apologies for the late reply.
I meant the `stat_prefix` down in this file (https://github.com/envoyproxy/envoy/blob/07f6286c2e9da40908f28022562ebd7ae332ca47/api/envoy/extensions/filters/network/rbac/v3/rbac.proto#L82). Isn't this for the specific RBAC filter?

I may be missing something, so feel free to correct me.
if the shadow rules have their own stat-prefix (shadow_rules_stat_prefix), and `stat_prefix` is set, why can't one  differentiate between the shadow vs non-shadow ones?

Comment (thomasvnoort) [reply api/envoy/extensions/filters/network/rbac/v3/rbac.proto:55]: Sorry, I misunderstood your question. I missed that the network RBAC filter already had a _configurable_ `stat_prefix`, unlike the http RBAC filter which inherits it from the connection manager and is not configurable.

So yes, you're right, for the network RBAC filter we don't need `rules_stat_prefix`. For the http RBAC filter we do since there's no way to namespace metrics from non-shadow rules for multiple filters currently. 

How would you like to proceed here? I can change the PR to only add the configuration for the http RBAC filter, but that would mean the network and http RBAC filter diverge a bit more.

Comment (adisuissa) [reply api/envoy/extensions/filters/network/rbac/v3/rbac.proto:55]: I think it may be beneficial to use the "same" solution in both the network and http filters (as long as it makes sense). Specifically, introduce `stat_prefix` to the http filter. WDYT?

Comment (thomasvnoort) [reply api/envoy/extensions/filters/network/rbac/v3/rbac.proto:55]: I agree, we should strive to use the same solution for both the network and http RBAC filter. 

If we make `stat_prefix` configurable on the http RBAC filter then it would override the value inherited from the connection manager (and default to that value if not defined, to not change current behaviour). That would work, but I think it breaks the expectation that `stat_prefix` is where a connection manager outputs all of its statistics; as mentioned on the [docs](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_conn_man/stats#statistics):

> Every connection manager has a statistics tree rooted at http.<stat_prefix>. with the following statistics:

If one would redefine `stat_prefix` for an http RBAC filter, those metrics would go to `http.my_rbac_filter.` whereas all connection manager metrics go to `http.my_connection_manager.`. That seems confusing to me and breaking the above promise. Moreover, there might also be other filters that inherit the `stat_prefix` in a similar fashion for emitting metrics, so that might mean the http RBAC filter would diverge there?

I agree that `rules_stat_prefix` is redundant for the network RBAC filter, but I think it's a minor issue at the benefit of having a similar interface for both and keeping HTTP metrics in a predictable namespace. Thoughts?

Comment (adisuissa) [reply api/envoy/extensions/filters/network/rbac/v3/rbac.proto:55]: Hmmm.... 2 observations:
1. this should be an opt-in config knob, that means that the stat-prefix will only be overridden if the field was set.
2. I'm not sure if it makes more sense to concatenate the name (something like `http.my_connection_manager.my_rbac_filter`).

I think that this should be similar to the what is being done for [shadow_rules_stat_prefix](https://github.com/envoyproxy/envoy/blob/deab45ffc8c7348518439b05fb41e2dca0d3ec39/api/envoy/extensions/filters/http/rbac/v3/rbac.proto#L64). I would suggest following the same pattern so the end result will be `<partA>.<shadow_rules_stat_prefix>.<partB>` for shadow rules and `<partA>.<rules_stat_prefix>.<partB>` for the non-shadow one. Let me know if you think this will work.

Comment (thomasvnoort) [reply api/envoy/extensions/filters/network/rbac/v3/rbac.proto:55]: The `shadow_rules_stat_prefix` on the http RBAC filter outputs to `http.<stat_prefix>.rbac.<shadow_rules_stat_prefix>.`. If we were to follow that for `rules_stat_prefix` we get `http.<stat_prefix>.rbac.<rules_stat_prefix>.`, which is exactly what this PR is currently doing.

But maybe I'm not entirely following what you're proposing for the http RBAC filter. Can you clarify what `partA` is here to be sure I understand you correctly?

I think we have three options here:
1. Only introduce `rules_stat_prefix` on the http RBAC filter (this causes network and http to diverge)
2. Make `stat_prefix` configurable on the http RBAC filter (this breaks the promise of where stats are sent, throwing multiple connection manager stats in one namespace)
3. Have `stat_prefix` and `rules_stat_prefix` on the network RBAC filter (some redundancy) per this PR

To me, the third option is most clear in terms of interface and expected behaviour. But, I might be missing a fourth option here.

Comment (adisuissa) [reply api/envoy/extensions/filters/network/rbac/v3/rbac.proto:55]: If the intended feature is already implemented in the network RBAC, then I don't think there is a need to modify it. Please correct me if that's not the case.

I see your point describing the differences between the network and http filters and I think there's no way to make both work the same which doesn't break current users (which we should avoid as much as possible).
For the HTTP RBAC, I think the new field can be added, I'm just wondering whether (1) it should replace the "http." prefix, or (2) appear after the "rbac." part. I'm more inclined to the 2nd option as it seems less disruptive to current users, and to will allow the hierarchical nature of the filters.

Comment (thomasvnoort) [reply api/envoy/extensions/filters/network/rbac/v3/rbac.proto]: Re the network RBAC filter. I agree so I removed those changes from this PR.

Re the http RBAC filter. I agree that the second option is preferred for the reasons you mentioned, plus that it aligns with how `shadow_rules_prefix` ends up in the metric namespace. This is how I originally implemented on this PR.

File: test/extensions/filters/http/rbac/rbac_filter_test.cc
Comment (adisuissa) [inline test/extensions/filters/http/rbac/rbac_filter_test.cc:243]: Instead of changing the stat_prefix for all tests, can there be a dedicated test where this behavior is validated (the motivation is that the rest of the tests are only targeting other validations, and that the case where "stat_prefix" isn't set is also exercised).

Also please change the value from "prefix_" to "rule_prefix_".

File: api/envoy/extensions/filters/http/rbac/v3/rbac.proto
Comment (adisuissa) [inline api/envoy/extensions/filters/http/rbac/v3/rbac.proto:38]: I may have missed that, but is there a test that validates that 2 RBAC filters have a different prefix for their stats?

**Review by adisuissa — COMMENTED** [review]
Thanks, overall LGTM. Left a few small comments.

File: api/envoy/extensions/filters/network/rbac/v3/rbac.proto
Comment (thomasvnoort) [reply api/envoy/extensions/filters/network/rbac/v3/rbac.proto]: Resolving this conversation as we agreed on not changing the network RBAC filter and adding the `rules_stat_prefix` only to the http RBAC filter.

File: docs/root/configuration/http/http_filters/rbac_filter.rst
Comment (thomasvnoort) [reply docs/root/configuration/http/http_filters/rbac_filter.rst]: See https://github.com/envoyproxy/envoy/pull/31835#discussion_r1487778639.

File: test/extensions/filters/http/rbac/rbac_filter_test.cc
Comment (thomasvnoort) [reply test/extensions/filters/http/rbac/rbac_filter_test.cc]: I added a separate test purely for `rules_stat_prefix` and left the existing `shadow_rules_stat_prefix` in tests alone (other than renaming the prefix from `prefix_` to `shadow_rules_prefix_`.

**Comment by kyessenov** [comment]
Let's make sure this and https://github.com/envoyproxy/envoy/pull/32313 are consistent.

**Comment by adisuissa** [comment]
> Let's make sure this and #32313 are consistent.

Good call. Let's make sure that a codeowner of RBAC reviews both PRs.
@yanavlasov are you a good person to assign this to, or can you nominate someone else?
Overall the current PR LGTM.

**Comment by thomasvnoort** [comment]
@adisuissa @yanavlasov Is there anything I can help with to get this PR merged? I'm aware of https://github.com/envoyproxy/envoy/pull/32313 since I know the author @henrymwang, so I can work with him to update his PR once mine is merged.

**Review by adisuissa — APPROVED** [review]
/lgtm api
LGTM, thanks!
/assign-from @envoyproxy/senior-maintainers

**Comment by alyssawilk** [comment]
@yanavlasov owns most of the rbac extensions so throwing over to him. @adisuissa bonus points for checking OWNERS.md/CODEOWNERS rather than doing random /assign as (IMO) it's worth having area experts where we have them.

**Comment by yanavlasov** [comment]
@kyessenov can you clarify what you are looking for in terms of consistency with #32313 , please? This change brigs feature parity between network and HTTP RBAC filters. It is different from #32313 I do not see why not have both.

/wait-any

**Comment by kyessenov** [comment]
@yanavlasov Yes, I was just trying to avoid a merge conflict - the custom prefix should be applied to per-policy stats I think.

**Comment by yanavlasov** [comment]
> @yanavlasov Yes, I was just trying to avoid a merge conflict - the custom prefix should be applied to per-policy stats I think.

Ok, noted

**Comment by yanavlasov** [comment]
@thomasvnoort please resolve merge conflict

/wait

File: source/extensions/filters/common/rbac/utility.h
Comment (yanavlasov) [inline source/extensions/filters/common/rbac/utility.h:37]: I would prefer to use absl::string_view here since you are changing this API anyway. Will it be too much trouble to change it?

**Review by yanavlasov — CHANGES_REQUESTED** [review]
LGTM from me with a small request.

Waiting for API review.

/wait-any

**Review by adisuissa — COMMENTED** [review]
/lgtm api

File: source/extensions/filters/common/rbac/utility.h
Comment (thomasvnoort) [reply source/extensions/filters/common/rbac/utility.h:37]: While I'd love to help out here, I'd prefer to merge this PR in its current form. I don't have the time right now to make this change and I'd rather avoid more conflicts, especially since https://github.com/envoyproxy/envoy/pull/32313 is related. Would it be ok to merge it as is?

**Review by yanavlasov — APPROVED** [review]

**Review by ggreenway — APPROVED** [review]
I had requested the addition of tag-extraction rules for the stats, which LGTM. I have not reviewed the rest of the PR in detail.

**Pull request merged**
